src/epoll.rs

```diff
@@ -196,22 +196,19 @@ impl Poller {
         #[cfg(target_os = "redox")]
         let timer_fd: Option<core::convert::Infallible> = None;
 
-        // Timeout in milliseconds for epoll.
-        let timeout_ms = match (timer_fd, timeout) {
-            (_, Some(t)) if t == Duration::from_secs(0) => 0,
-            (None, Some(t)) => {
-                // Round up to a whole millisecond.
-                let mut ms = t.as_millis().try_into().unwrap_or(i32::MAX);
-                if Duration::from_millis(ms as u64) < t {
-                    ms = ms.saturating_add(1);
-                }
-                ms
-            }
-            _ => -1,
+        // Timeout for epoll.
+        let timeout = match (timer_fd, timeout) {
+            (_, Some(t)) if t == Duration::from_secs(0) => Some(Timespec::default()),
+            (None, Some(t)) => Some(Timespec::try_from(t).map_err(|_| Errno::INVAL)?),
```

Comment:
This will return an error in cases that did not previously return an error, right?

Reply:
This will produce an error if the seconds part of "t" doesn't fit into `i64`.

Reply:
Note how big `2^63 - 1` (the largest number representable using `i64`) is: https://www.wolframalpha.com/input?i=2%5E63+-+1+seconds

It is not useful to support larger values, at least not for timeouts.

Reply:
It doesn't matter how “unlikely it is to be reached by natural increase”; people rather use such values because of their nature.

See the issues linked from the issues I linked to, e.g. https://github.com/crossbeam-rs/crossbeam/issues/795.